### PR TITLE
Option to spatially partition tiles before tiles are read.

### PR DESCRIPTION
**Pull request by metasim**

### CODE REVIEW

**Comment by metasim** [comment]
Note: `SFCIndexerSpec.scala` went from 34 minutes to 12 minutes with this change.

File: docs/src/main/paradox/release-notes.md [hunk 1]
```diff
@@ -17,7 +23,7 @@
 * Updated to GeoTrellis 2.3.3 and Proj4j 1.1.0.
 * Fixed issues with `LazyLogger` and shading assemblies ([#293](https://github.com/locationtech/rasterframes/issues/293))
 * Updated `rf_crs` to accept string columns containing CRS specifications. ([#366](https://github.com/locationtech/rasterframes/issues/366))
-* Added `rf_spatial_index` function. ([#368](https://github.com/locationtech/rasterframes/issues/368))
+* Added `rf_xz2_index` function. ([#368](https://github.com/locationtech/rasterframes/issues/368))
```
Comment (vpipkt) [inline docs/src/main/paradox/release-notes.md]: i think you should not change this and instaead in 0.8.5 notes, say "renamed rf_spatial_index function to rf_xz2_index"

File: docs/src/main/paradox/reference.md [hunk 2]
```diff
@@ -67,14 +67,22 @@ See also GeoMesa [st_envelope](https://www.geomesa.org/documentation/user/spark/
...
     
 Constructs a XZ2 index in WGS84/EPSG:4326 from either a Geometry, Extent, ProjectedRasterTile and its CRS. This function is useful for [range partitioning](http://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=registerjava#pyspark.sql.DataFrame.repartitionByRange).
 
+### rf_z2_index
+
+    Long rf_z2_index(Geometry geom, CRS crs)
+    Long rf_z2_index(Extent extent, CRS crs)
+    Long rf_z2_index(ProjectedRasterTile proj_raster)
+    
+Constructs a Z2 index in WGS84/EPSG:4326 from either a Geometry, Extent, ProjectedRasterTile and its CRS. First the native extent is extracted or computed, and then center is used as the indexing location. This function is useful for [range partitioning](http://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=registerjava#pyspark.sql.DataFrame.repartitionByRange).
```
Comment (vpipkt) [inline docs/src/main/paradox/reference.md]: please link to raster read section about indexing.

File: docs/src/main/paradox/reference.md [hunk 3]
```diff
@@ -67,14 +67,22 @@ See also GeoMesa [st_envelope](https://www.geomesa.org/documentation/user/spark/
...
 
-### rf_spatial_index
+### rf_xz2_index
 
-    Long rf_spatial_index(Geometry geom, CRS crs)
-    Long rf_spatial_index(Extent extent, CRS crs)
-    Long rf_spatial_index(ProjectedRasterTile proj_raster, CRS crs)
+    Long rf_xz2_index(Geometry geom, CRS crs)
+    Long rf_xz2_index(Extent extent, CRS crs)
+    Long rf_xz2_index(ProjectedRasterTile proj_raster)
```
Comment (vpipkt) [inline docs/src/main/paradox/reference.md]: below we should mention and link to the raster read section about spatial index parititoning.

File: datasource/src/main/scala/org/locationtech/rasterframes/datasource/raster/RasterSourceRelation.scala [hunk 4]
```diff
@@ -40,13 +41,16 @@ import org.locationtech.rasterframes.tiles.ProjectedRasterTile
   * @param catalogTable Specification of raster path sources
   * @param bandIndexes band indexes to fetch
   * @param subtileDims how big to tile/subdivide rasters info
+  * @param lazyTiles if true, creates a lazy representation of tile instead of fetching contents.
+  * @param spatialIndexPartitions if not None, adds a spatial index. If Option value < 1, uses the value of `numShufflePartitions` in SparkContext.
```
Comment (vpipkt) [inline datasource/src/main/scala/org/locationtech/rasterframes/datasource/raster/RasterSourceRelation.scala]: if the value is >=1 does it become the number of partitions used? that is not exactly clear from a quick read. could epxand this

File: pyrasterframes/src/main/python/pyrasterframes/__init__.py [hunk 5]
```diff
@@ -146,6 +149,16 @@ def temp_name():
     if band_indexes is None:
         band_indexes = [0]
 
+    if spatial_index_partitions is False:
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/pyrasterframes/__init__.py]: I propose the below which covers some additional cases e.g. user passing in a literal `int` 0.

Note that when `a is None` then a logical test on `a` returns False. 

```suggestion
    if spatial_index_partitions:
        if int(spatial_index_partitions) < 1:
            spatial_index_partitions = '-1'
       
        options.update({"spatial_index_partitions": str(int(spatial_index_partitions))})
```

File: pyrasterframes/src/main/python/tests/RasterSourceTest.py [hunk 6]
```diff
@@ -211,3 +211,9 @@ def test_catalog_named_arg(self):
         df = self.spark.read.raster(catalog=self.path_pandas_df(), catalog_col_names=['b1', 'b2', 'b3'])
         self.assertEqual(len(df.columns), 7)  # three path cols, three tile cols, and geo
         self.assertTrue(df.select('b1_path').distinct().count() == 3)
+
+    def test_spatial_partitioning(self):
+        df = self.spark.read.raster(self.path(1, 1), spatial_index_partitions=True)
+        self.assertTrue('spatial_index' in df.columns)
+        # Other tests?
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/tests/RasterSourceTest.py]: maybe try passing in the following values with the expected results

int(0) -> no spatial index
float(99) -> df.rdd.getNumPartitions == 99
str("42.0") -> df.rdd.getNumPartitions == 42
False -> no spatial index

File: pyrasterframes/src/main/python/tests/VectorTypesTests.py [hunk 7]
```diff
@@ -157,15 +157,39 @@ def test_geojson(self):
...
+        # Test against proj_raster (has CRS and Extent embedded).
+        df = self.spark.read.raster(self.img_uri)
+        result_one_arg = df.select(rf_xz2_index('proj_raster').alias('ix')) \
+            .agg(F_min('ix')).first()[0]
+
+        result_two_arg = df.select(rf_xz2_index(rf_extent('proj_raster'), rf_crs('proj_raster')).alias('ix')) \
+            .agg(F_min('ix')).first()[0]
+
+        self.assertEqual(result_two_arg, result_one_arg)
+        self.assertEqual(result_one_arg, 55179438768)  # this is a bit more fragile but less important
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/tests/VectorTypesTests.py]: 👍

File: pyrasterframes/src/main/python/tests/VectorTypesTests.py [hunk 8]
```diff
@@ -157,15 +157,39 @@ def test_geojson(self):
...
+
         # Custom resolution
-        df = self.df.select(rf_spatial_index(self.df.poly_geom, rf_crs(lit("EPSG:4326")), 3).alias('index'))
+        df = self.df.select(rf_xz2_index(self.df.poly_geom, rf_crs(lit("EPSG:4326")), 3).alias('index'))
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/tests/VectorTypesTests.py]: 😹

File: docs/src/main/paradox/release-notes.md (hunk 1 above)
Comment (metasim) [reply docs/src/main/paradox/release-notes.md]: That was an accidental rename. 🤦‍♂

**Pull request merged**